### PR TITLE
[Driver] Fix FPGAEmulationMode naming confusion

### PR DESCRIPTION
**Pull request by srividya-sundaram**
Fixes https://github.com/intel/llvm/issues/4061

### CODE REVIEW

**Comment by srividya-sundaram** [comment]
Failures seem related to `sycl::multi_ptr` and is un-related to changes in this PR.
Other PRs are also failing with the same errors related to `sycl::multi_ptr.`

File: clang/include/clang/Driver/Driver.h [hunk 1]
```diff
@@ -718,6 +718,28 @@ class Driver {
...
+    FPGAEmulationMode
+  } OffloadCompileMode = FPGAEmulationMode;
+
+  bool IsFPGAHWMode() const { return OffloadCompileMode == FPGAHWMode; }
+
+  bool IsFPGAEmulationMode() const {
+    return OffloadCompileMode == FPGAEmulationMode;
+  }
+
+  void setOffloadCompileMode(StringRef ModeVale) {
```
Comment (mdtoguchi) [inline clang/include/clang/Driver/Driver.h:734]: Use of a string to set the mode feels off, and a bit arbitrary.  Why not just set with the enumerated value?

File: clang/include/clang/Driver/Driver.h [hunk 2]
```diff
@@ -718,6 +718,28 @@ class Driver {
     return IsOffload ? OffloadLTOMode : LTOMode;
   }
 
+  /// FPGA Emulation Mode.  By default, this is true due to the fact that
+  /// an external option setting is required to target hardware.
+  enum DeviceMode {
+    FPGAHWMode,
+    FPGAEmulationMode
+  } OffloadCompileMode = FPGAEmulationMode;
```
Comment (mdtoguchi) [inline clang/include/clang/Driver/Driver.h:726]: We should extend this a bit more.  By default, there shouldn't be a specific mode set, so we can introduce some default 'unset' value.  For example, if one were to use `'-fsycl-targets=spir64_x86_64` this device target has nothing to do with FPGA, but we would have `FPGAEmulationMode` set anyway.  Having an 'unset' type of default value can give use that extra bit of control.

Comment (srividya-sundaram) [reply clang/include/clang/Driver/Driver.h:726]: `FPGAEmulationMode` is set as the default value in the old code. It is not being set from anywhere else.
If we were to remove the default `FPGAEmulationMode` value, where should we set this value from?

Comment (mdtoguchi) [reply clang/include/clang/Driver/Driver.h:726]: The default would be a new value, maybe `UnsetDeviceMode` or something like that.

**Review by mdtoguchi — COMMENTED** [review]
This looks like a good opportunity to update around https://github.com/intel/llvm/blob/2fc89acc631da8567bae1dfdf3013b9ae79ec4d1/clang/lib/Driver/Driver.cpp#L6417 to use the enum - what do you think?

**Comment by srividya-sundaram** [comment]
> This looks like a good opportunity to update around
> 
> https://github.com/intel/llvm/blob/2fc89acc631da8567bae1dfdf3013b9ae79ec4d1/clang/lib/Driver/Driver.cpp#L6417
> 
> to use the enum - what do you think?

Could you please elaborate?
[IsFPGAEmulation](https://github.com/intel/llvm/blob/2fc89acc631da8567bae1dfdf3013b9ae79ec4d1/clang/lib/Driver/Driver.cpp#L6394) is already invoking `IsFPGAEmulationMode` to get the enum value. Did you mean something else?

**Comment by mdtoguchi** [comment]
> > This looks like a good opportunity to update around
> > https://github.com/intel/llvm/blob/2fc89acc631da8567bae1dfdf3013b9ae79ec4d1/clang/lib/Driver/Driver.cpp#L6417
> > 
> > to use the enum - what do you think?
> 
> Could you please elaborate? [IsFPGAEmulation](https://github.com/intel/llvm/blob/2fc89acc631da8567bae1dfdf3013b9ae79ec4d1/clang/lib/Driver/Driver.cpp#L6394) is already invoking `IsFPGAEmulationMode` to get the enum value. Did you mean something else?

In the array of pairs, you can see that the `IsFPGAEmulation` bool is just checking against the second value of the pair to see if the type is an emulation type or not.  Instead of doing that, just do an IsFPGAHardware or IsFPGAEmulation kind of check.  Just a suggestion though, wanted to see what your thoughts were.

File: clang/lib/Driver/Driver.cpp [hunk 3]
```diff
@@ -6416,12 +6419,13 @@ class OffloadingActionBuilder final {
         {types::TY_FPGA_AOCR_EMU, true}};
     for (const auto &ArchiveType : FPGAAOCTypes) {
       bool BinaryFound = hasFPGABinary(C, InputName, ArchiveType.first);
-      if (BinaryFound && ArchiveType.second == IsFPGAEmulation) {
+      if (BinaryFound && ArchiveType.second == D.IsFPGAEmulationMode()) {
```
Comment (mdtoguchi) [inline clang/lib/Driver/Driver.cpp:6422]: With these changes - the boolean variable is no longer needed from above.

**Review by mdtoguchi — APPROVED** [review]
LGTM - thanks

**Comment by srividya-sundaram** [comment]
@intel/llvm-gatekeepers Hi, this is ready for merge. Failures reported are un-related to changes in this PR.

**Comment by steffenlarsen** [comment]
> @intel/llvm-gatekeepers Hi, this is ready for merge. Failures reported are un-related to changes in this PR.

@srividya-sundaram - Please resolve the merge conflict.

**Pull request merged**